core/src/main/scala/tech/beshu/ror/accesscontrol/factory/decoders/rules/auth/LdapRulesDecoders.scala

```diff
@@ -109,22 +109,40 @@ class LdapAuthorizationRuleDecoder(ldapDefinitions: Definitions[LdapService],
       .instance { c =>
         for {
           name <- c.downField("name").as[LdapService.Name]
-          groupsAnd <- c.downField("groups_and").as[Option[GroupsLogic.And]]
-          groupsOr <- c.downFields("groups_or", "groups").as[Option[GroupsLogic.Or]]
+          groupsAnd <- c.downFields("groups_and", "groups_all_of").as[Option[GroupsLogic.And]]
+          groupsOr <- c.downFields("groups_or", "groups", "groups_any_of").as[Option[GroupsLogic.Or]]
+          groupsNotAllOf <- c.downFields("groups_not_all_of").as[Option[GroupsLogic.NotAllOf]]
+          groupsNotAnyOf <- c.downFields("groups_not_any_of").as[Option[GroupsLogic.NotAnyOf]]
           ttl <- c.downFields("cache_ttl_in_sec", "cache_ttl").as[Option[PositiveFiniteDuration]]
-        } yield (name, ttl, groupsOr, groupsAnd)
+        } yield (name, ttl, groupsOr, groupsAnd, groupsNotAllOf, groupsNotAnyOf)
       }
       .toSyncDecoder
       .mapError(RulesLevelCreationError.apply)
       .emapE {
-        case (_, _, None, None) =>
+        case (_, _, None, None, None, None) =>
           Left(RulesLevelCreationError(Message(errorMsgNoGroupsList(LdapAuthorizationRule.Name))))
-        case (_, _, Some(_), Some(_)) =>
-          Left(RulesLevelCreationError(Message(errorMsgOnlyOneGroupsList(LdapAuthorizationRule.Name))))
-        case (name, ttl, Some(groupsOr), None) =>
+        case (name, ttl, Some(groupsOr), None, None, None) =>
           createLdapAuthorizationRule(name, ttl, groupsOr, ldapDefinitions)
-        case (name, ttl, None, Some(groupsAnd)) =>
+        case (name, ttl, None, Some(groupsAnd), None, None) =>
```

Comment:
oh, it's getting complicated. Are we able to somehow extract the groups logic parsing to a separate, private method?

Reply:
Done. It was possible to extract common decoder for GroupsLogic for both Authorization and Auth decoders